tests/ui/abi/compatibility.rs

```diff
@@ -277,6 +277,10 @@ test_abi_compatible!(zst_unit, Zst, ());
 test_abi_compatible!(zst_array, Zst, [u8; 0]);
 test_abi_compatible!(nonzero_int, NonZero<i32>, i32);
 
+// `#[repr(C)]` enums should not change ABI based on individual variant inhabitedness.
```

Comment:
```suggestion
// `#[repr(C)]` enums should not change ABI based on individual variant inhabitedness.
// (However, this is *not* a guarantee.)
```
ABI compatibility is only guaranteed for very few cases, see [here](https://doc.rust-lang.org/stable/std/primitive.fn.html#abi-compatibility). I don't think we want to guarantee it for this case.

We want to guarantee that they have the same layout, but you are testing a lot more than that here.

Reply:
It would be nice to have something like `assert_abi_compatible!()` available to the library authors. While this property is not _guaranteed_, it would probably hold for most platforms. Having a means to fail compilation if it doesn’t hold for the target platform would be a useful safeguard.